x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java

```diff
@@ -267,13 +277,17 @@ synchronized void stopProcess() {
             try {
                 stateStreamer.cancel();
                 process.get().kill(true);
+                processContextByAllocation.remove(taskId);
```

Comment:
instead of removing here remove it in `stopDeployment(TrainedModelDeploymentTask task)`  line 170
otherwise another thread could come in and see that the process context is still in the map and try to infer on while it is stopping

Reply:
I am not sure about this. It may be that `stopped` actually failed. So the process is still around and never exited. If we remove it from the map without making sure that its actually stopped, we lose access to it.

I do agree that when the `task.isStopped()` infer should not work. I added a clause there and shored up stopping the task itself.